### ISSUE TITLE
(4.6.2 -> 4.6.3 regression) Calling .set on a subfield of a deselected subdoc wipes out other fields on the subdoc

### ISSUE THREAD

**Issue by n-rook** [Automattic/mongoose#4651]
Suppose I have a schema defining a single embedded subdocument. I don't select this subdocument when I fetch the parent document, but I set a field on it anyway using .set.

In 4.6.2, this works as expected: the field I set changes, and the rest of the subdocument remains unchanged.

In 4.6.3, this wipes out the rest of the subdocument's fields, and gives it a new ID.

This feels like a bug, though it may be intended behavior.

```javascript
const mongoose = require('mongoose');
const Promise = require('bluebird');

mongoose.Promise = Promise;
const connection = mongoose.connect('mongodb://localhost/minimal', {});
mongoose.set('debug', true);

const ballSchema = mongoose.Schema({
  kind: String,
  mass: Number
});

const sportSchema = mongoose.Schema({
  ball: ballSchema
});

const Sport = mongoose.model('Sport', sportSchema);

new Sport({ball: {kind: 'birdie', mass: 4.75 }}).save()
  .then((sport) => Sport.findById(sport._id).select('-ball'))
  .then((sport) => {
    sport.set('ball.kind', 'shuttlecock');

    // On 4.6.2, this save issues this update:
    // sports.update({ _id: ObjectId(...) },
    //   { '$set': { 'ball.kind': 'shuttlecock' } })
    // On 4.6.3, it issues this update:
    // sports.update({ _id: ObjectId("...") },
    //   { '$set': { ball: { kind: 'shuttlecock', _id: ObjectId("...") } } })
    return sport.save();
  })
  .then((sport) => Sport.findById(sport._id).select('ball'))
  .then((sport) => {
    // On 4.6.2, sport.ball.mass === 4.7.5
    // On 4.6.3, sport.ball.mass is undefined
    console.log('sport.ball.mass', sport.ball.mass);
  });
```

**Comment by vkarpov15** [comment]
Thanks for the detailed repro, will investigate

**Comment by vkarpov15** [comment]
Yeah this is related to the fix for #4578. I would argue this is expected behavior, because when you do `.set('ball.kind')` you're implicitly creating a new `ball`, which in turn gets tracked as setting `ball` to `{ kind: 'shuttlecock' }`. What's your use case for this behavior?

**Comment by n-rook** [comment]
I think it's unexpected behavior, but I agree it's ambiguous. It's not obviously a bug.

My actual use case was as follows: I have a CRUD app. On the schema for my primary type of object, I have a `notification` field. `notification` is an object containing these fields:

- `notification.pending` is true if we need to email out more notifications.
- I have some other fields for the last time we sent out notifications, which users have seen the document lately, etc.

I had a pre-save hook setting `notification.pending` to true on each save. This happened even if `notification` wasn't selected. But then we updated to 4.6.3, and this hook began clobbering all the other fields on `notification` if the saved document didn't select it.

I wound up working around this just by always selecting `notification`, and editing my pre-save hook to complain if `notification` isn't selected.

**Comment by vkarpov15** [comment]
Yeah I can see that's a problem. I think the real issue here though is that mongoose is saving modifications to unselected fields, which seems like a gnarly gotcha. We'll consider changing that for 5.0, what do you think?

**Comment by n-rook** [comment]
Yeah, that makes sense to me. Thanks for taking a look!

**Comment by dietergeerts** [comment]
Does this mean that the general idea is that only selected paths can be changed? So there is no option to pass to `set` to update even if it's not selected? I'm just encountering an issue with this that I though should work: https://github.com/Automattic/mongoose/issues/5800#issuecomment-692140093. Though I understand that this is a difficult issue/use-case.

**Comment by vkarpov15** [comment]
@dietergeerts I'd argue that yes, `save()` should only pick up selected paths. I opened up a new issue #9427 to track your comments from #5800.
